connector/jobrunner/channels.py

```diff
@@ -0,0 +1,629 @@
+
+        If the channel does not exist it is created.
+        The configuration is applied on the channel before returning it.
+        If some of the parent channels are missing when creating a subchannel,
+        the parent channels are auto created with an infinite capacity
+        (except for the root channel, which defaults to a capacity of 1
+        when not configured explicity).
+        """
+        channel = self.get_channel_by_name(config['name'], autocreate=True)
+        channel.configure(config)
+        return channel
+
+    def get_channel_by_name(self, channel_name, autocreate=False):
+        """Return a Channel object by its name.
+
+        If it does not exist and autocreate is True, it is created
+        with a default configuration and inserted in the Channels structure.
+        If autocreate is False and the channel does not exist, an exception
+        is raised.
+
+        >>> cm = ChannelManager()
+        >>> c = cm.get_channel_by_name('root', autocreate=False)
+        >>> c.name
+        'root'
+        >>> c.fullname
+        'root'
+        >>> c = cm.get_channel_by_name('root.sub', autocreate=True)
+        >>> c.name
+        'sub'
+        >>> c.fullname
+        'root.sub'
+        >>> c = cm.get_channel_by_name('sub', autocreate=True)
+        >>> c.name
+        'sub'
+        >>> c.fullname
+        'root.sub'
+        >>> c = cm.get_channel_by_name('autosub.sub', autocreate=True)
+        >>> c.name
+        'sub'
+        >>> c.fullname
+        'root.autosub.sub'
+        """
+        if channel_name in self._channels_by_name:
+            return self._channels_by_name[channel_name]
+        if not autocreate:
+            raise ChannelNotFound('Channel %s not found' % channel_name)
+        parent = self._root_channel
+        for subchannel_name in channel_name.split('.'):
+            if subchannel_name == 'root':
+                continue
+            subchannel = parent.get_subchannel_by_name(subchannel_name)
+            if not subchannel:
+                subchannel = Channel(subchannel_name, parent, capacity=None)
+                self._channels_by_name[subchannel.fullname] = subchannel
+            parent = subchannel
+        return parent
+
+    def notify(self, db_name, channel_name, uuid,
+               seq, date_created, priority, eta, state):
+        job = self._jobs_by_uuid.get(uuid)
+        if not job:
+            if not channel_name:
+                channel = self._root_channel
+            else:
+                try:
+                    channel = self.get_channel_by_name(channel_name)
+                except ChannelNotFound:
+                    _logger.warning('unknown channel %s, '
+                                    'using root channel for job %s',
+                                    channel_name, uuid)
+                    channel = self._root_channel
+            job = ChannelJob(db_name, channel, uuid,
+                             seq, date_created, priority, eta)
```

Comment:
Be careful. In Odoo python `date`, `datetime` and `time`  are returned as `string` by psycopg2. (see https://github.com/odoo/odoo/blob/8.0/openerp/sql_db.py#L61 and  http://initd.org/psycopg/docs/extensions.html#psycopg2.extensions.new_type). If you assign a `string` as `eta` to your ChannelJob you'll get a `Comparison error between str and datetime` when popping you job at line #L418
```python
    while not self.capacity or len(self._running) < self.capacity:
            job = self._queue.pop(now=datetime.now())
```
due to #L275
```python
    def pop(self, now):
        if len(self._eta_queue) and self._eta_queue[0].eta <= now:
```